### ISSUE TITLE
[MNT] Test consistency between CPU & GPU version of ROCKET

### ISSUE THREAD

**Issue by baraline** [aeon-toolkit/aeon#1248]
### Describe the issue

After the introduction of the GPU version of ROCKET in #1199, we need new test in order to make sure the CPU and GPU version of rocket are consistent when given the same random seed.

### Suggest a potential alternative/fix

The goal of this issue is to implement a test [here](https://github.com/aeon-toolkit/aeon/tree/main/aeon/transformations/collection/convolution_based/rocketGPU/tests) that will transform a small test dataset using both CPU and GPU versions of ROCKET initialized with the same random seed, and test that the transformation result are similar using the `assert_array_almost_equal` function from `numpy.testing`

**Comment by hadifawaz1999** [comment]
given both versions uses numpy to randomly generate the convolution parameters, it should be simple, if it was one numpy one tensorflow than it would have been problematic

we should make sure both versions are using the new api generator in numpy, use the same dataset, and make sure we both normalize the input

NOTE: the rocket-gpu only works on equal length series, limitations of accelerating with gpu is batching, and batching only works with equal length series
So we should test only with equal length series, univariate and multivariate

**Comment by TonyBagnall** [comment]
replaced by #1699

**Comment by Adityakushwaha2006** [comment]
Hey, I would like to work on this issue, I've reviewed the previous changes and understood the issue.
I believe I can prop up the test cases and a possible fix for the same.

**Comment by Adityakushwaha2006** [comment]
@aeon-actions-bot assign @Adityakushwaha2006

**Comment by Adityakushwaha2006** [comment]
Hey moderators, I’ve been digging into the Rocket (CPU) vs ROCKETGPU implementations to figure out why exactly they diverge.

It turns out it’s a bit deeper than just the Random Number Generator. There are two structural mismatches we need to handle:

**Sparse vs Dense: The CPU version randomly selects a subset of channels for each kernel (Sparse behavior). The current GPU version convolves over all channels (Dense behavior).

Padding differences: CPU uses explicit integer padding logic, whereas the GPU relies on TensorFlow's "SAME"/"VALID" strings, which handles edge cases differently.**

The Fix I'm Proposing: To get them consistent without rewriting the entire GPU kernel, I think we should just bridge them:

Unified Gen: Inside ROCKETGPU.fit, we call the standard CPU _generate_kernels to get the "source of truth" weights.

Adaptation: We take those sparse kernels and convert them to dense tensors by zero-padding the unused channels. This mathematically replicates the sparse behavior on the GPU.

Explicit Padding: We calculate the exact padding integers on the CPU side and pass them to TF (using ZeroPadding1D or similar) to avoid the fuzzy behavior of "SAME".

This guarantees that Rocket(seed=42) and ROCKETGPU(seed=42) are doing the exact same math.

Does this approach sound good to you? If so, I can start implementing that bridge logic. Please let me know  ,Thanks : )

**Comment by hadifawaz1999** [comment]
Hello, thank you for taking some time into looking at this issue @Adityakushwaha2006 ! ever since i got rocketGPU in aeon i have been meaning to fix the channels issue you mentioned. However even on univariate cases it still has issues and it probably is from random number generators

Let me know when you start working on this i can take a look at the PR !

**Comment by Adityakushwaha2006** [comment]
@hadifawaz1999 
Thanks for the pointer, that aligns exactly with my debugging.

I traced the specific divergence to how the kernels are initialized. The CPU implementation relies on np.random (passed into Numba), while the GPU version uses tf.random. Even with the same seed, the kernel weight distributions drift because the generation logic isn't strictly synchronized across the backends.

I have a two-step plan to fix this:

Step 1 (Immediate): I am about to open a PR that unifies the Rocket interface (adding a device="gpu" parameter). This structural refactor makes it trivial to run side-by-side verification tests between CPU and GPU without code duplication. I have already made significant progress on this issue. ( issue listed : #3174 )

Step 2 (main step for #1248) : Once that interface is merged, I will submit the math fix to synchronize the random states and resolve the channel inconsistencies you mentioned. Ill refine my approach and let you know of any significant advancements in these as in when they come up.

I'll tag you on the Step 1 PR shortly.

**Comment by Adityakushwaha2006** [comment]
@hadifawaz1999 @MatthewMiddlehurst

**Update:** I have completed the comprehensive analysis of the CPU vs. GPU ROCKET implementations across 3 datasets (GunPoint, BasicMotions, ECG200).

The Findings are critical: We have 0% reproducibility between CPU and GPU ROCKET, with massive divergence in multivariate cases.

**Key Data Points:**

- Hash Match Rate: 0% (All datasets)
- BasicMotions (Multivariate): 2,050% difference in specific feature values. The CPU uses random channel subsets, while the GPU implementation currently forces all channels, leading to fundamentally different kernels.
- GunPoint (Univariate): ~9.6% mean difference (due to RNG: MT19937 vs PCG64).

**Root Cause Identified:**

- RNG Mismatch: NumPy (CPU) uses Mersenne Twister; TensorFlow (GPU) uses PCG64.
- Algorithm Divergence: The GPU implementation does not implement the random channel selection logic present in the CPU version.
(Ill dig deeper into the causes in the trailing message)

Proposed Action Plan (Starting Now): I am moving to the implementation phase to enforce parity.

Unified RNG: I will force the GPU class to generate weights/indices using numpy (CPU) first, then inject them into the TensorFlow graph.

Channel Parity: I will implement the missing "random channel subset" logic in the GPU version.

Backward Compatibility: I plan to add a legacy_rng=False flag to ROCKETGPU to prevent breaking existing users' experiments.

I'm proceeding with this implementation on a fresh branch. Let me know if you have objections to the legacy flag approach.

**Comment by Adityakushwaha2006** [comment]
The Main points of pain are:

1. **Channel Selection** : In the standard CPU version, ROCKET uses a "Sparse Channel" logic, meaning for every kernel, it randomly selects a subset of channels (e.g., just 3 out of 6) to look at. The GPU version, however, doesn't do this. It currently convolves over ALL channels for every single kernel. So, for the exact same seed, the GPU is mathematically seeing "more" data per kernel than the CPU. This changes the resulting feature map entirely, which explains why BasicMotions (6 channels) was so far off.

2. **The RNG** :  Different Even if we fix the logic, we have a seed problem. The CPU implementation relies on the legacy np.random.seed() (which uses the Mersenne Twister 1997 algo), while the GPU code instantiates the newer np.random.default_rng() (which uses PCG64). So even with seed=42 on both, the sequence of random numbers they generate is completely uncorrelated. 

3. **Padding mismatches** : The CPU code calculates specific integer padding values based on the dilation and kernel size. The GPU version relies on TensorFlow's high-level "SAME" or "VALID" string modes. TensorFlow's "SAME" logic is a heuristic—it doesn't always align 1:1 with the explicit integer padding used in the Numba implementation, especially for edge cases in time-series length.


Supporting Data (helps interpretation):

<img width="527" height="599" alt="Image" src="https://github.com/user-attachments/assets/7f15988d-4cd8-46a6-8d0f-c8fae884af13" />

<img width="607" height="474" alt="Image" src="https://github.com/user-attachments/assets/57db0059-52f5-4586-a5c7-1a929c8059b7" />

(Environment used : CPU: Intel Xeon, GPU: Tesla T4)

**Comment by Adityakushwaha2006** [comment]
@hadifawaz1999 I needed some guidance here. I wanted to consult with you before implementing further changes, as the next steps might qualitatively shift the GPU version's capabilities.

As you might have read, I have achieved almost perfect parity for the kernel generation, so the fit operation is essentially following the same logic and creating the same RNG-to-Kernel values.

However, the problem now falls on the transformation step, as stated here: https://github.com/aeon-toolkit/aeon/pull/3177#issuecomment-3671642393

- Univariate datasets (1 channel): 0.3-0.5% divergence
- Multivariate dataset (6 channels): 244% divergence
The problem appears specific to multi-channel handling.

I have been analyzing the issue, and currently, as per my understanding:

- It is not channel ordering (CPU and GPU select identical channels in identical order).
- It is not padding (output lengths match perfectly).
- It is not kernel weights (weights match exactly).
- It likely is the convolution computation itself (tf.nn.conv1d producing different results than the CPU manual loop).

### Currently, these are the actions we can take:
**Option 1: Accept Small Divergence (0.3-0.5%)**
Action: Only fix the BasicMotions extreme case.
Pros: Quick resolution; maintains GPU performance.

**Option 2: Force Exact CPU-GPU Parity**
Action: Replace tf.nn.conv1d with a manual implementation.
Approaches:
A) Python loops: Simple but SLOW (defeats GPU purpose).
B) Numba JIT: Faster but still not GPU-accelerated.
C) Custom TF operation: Complex, requires TF expertise.
Pros: Perfect reproducibility; matches CPU exactly.
Cons: Loses GPU performance benefits; may defeat ROCKETGPU's purpose.

**Option 3: Hybrid Approach**
Action: Keep TensorFlow for univariate (low divergence) and use an alternative for multivariate (high divergence).
Pros: Best performance where it works; fixes the critical BasicMotions case.
Cons: Maintains two different code paths.

**Option 4: Investigate TF Conv1d Semantics**
Action: Deep dive into why tf.nn.conv1d differs.
Pros: May find a simple fix; keeps GPU benefits.
Cons: Uncertain timeline; may not yield a solution.

I am ready to provide any additional details required for this decision, as I have the benchmarking and validation scripts in place and running.

Please let me know how you would like to proceed and what the primary expectation is for this fix.

**Comment by Adityakushwaha2006** [comment]
Since my previous message, I conducted comprehensive testing across all available multivariate datasets to determine if BasicMotions was an isolated case or indicative of a systemic issue.

### Test 1: All Multivariate Datasets (24 tested)
Distribution of divergence:

- LOW (<1%): 4 datasets (17%)
- MEDIUM (1-10%): 2 datasets (8%)
- HIGH (>10%): 18 datasets (75%)** 

### Test 2: Pattern Analysis
Hypothesis: Short timepoint sequences lead to High divergence
Findings:
LSST (36 timepoints) - 23,091% 
Cricket (1197 timepoints) - 0.04% 
(However: Pattern is NOT reliable)
EthanolConcentration (1751 timepoints) - complete divergence
Some short sequences still have low divergence
**Conclusion: Timepoints correlate weakly but are not the primary factor.**

Also , increasing sample size makes divergence worse.




### Current Pattern identified:

- **Univariate (1 channel): GPU works well (0.3-0.5% divergence)**
- **Multivariate (2+ channels): GPU produces large errors (complete divergence in 75% of cases)**

(The aforementioned issue of even univariate cases diverging , is solved to the maximum , there is practically no divergence on their kernel creation and whatever divergence (minimal) is being created , that is due to the tf.convolution algorithm. However its managable)

In conclusion , the multivariate divergences are systemic , and cannot be tackled unless we deep dive into a complete architectural revamp for the convolution step / change the methodology to match it with CPU variant.

Im awaiting your response for this situation.
_Currently , what i see personally should be our approach , is to fix the univariate cases and push this solution (as this fixes univariate and intermediate kernel parity) , and then maybe continue discussion/research on the transform step convolution matching._

**Comment by MatthewMiddlehurst** [comment]
Thanks for looking into the problem. After a quick look, not a fan of the rework. The only place there should be major deviation in the current implementation should be in generating kernels right? No way doing that is more efficient with a GPU given its essentially just randomly generating a few numbers.

**Comment by baraline** [comment]
My thinking was that the GPU part would only help in computing the convolutions, not generating the kernels as you said. Although CuPy seem to be a great tool for those kinds of numpy to GPU implementation !

**Comment by Adityakushwaha2006** [comment]
@MatthewMiddlehurst, in the original GPU implementation (currently in main) - no kernel parity, no feature parity. Even univariate datasets diverge.
In PR #3177 ( my TensorFlow attempt), I achieved kernel parity (<1e-7) , but the transform convolution step still created divergence due to TF convolution behavior. Univariate datasets showed parity, but multivariate were still divergent.

Thats the reason I wrote the custom CUDA kernel in the CuPy version...to fix the 
transform step, not kernel generation. ( i can reopen #3177 if you want to review the TensorFlow parity data.)

**Comment by Adityakushwaha2006** [comment]
@baraline , thats the exact approach i had followed in the CuPy implementation. Matthew said that he'd like other maintainers to have a look on the CuPy implementation before making a decision. 
Since you seem to be familiar with CuPy , could you have a look at #3211 , whenever you get the chance. 
Thanks : )

**Comment by Adityakushwaha2006** [comment]
The approach for the TF implementation had gotten really messy, thereby I closed it, Ill start a fresh PR with keeping all our discussions in mind.
Taking a break from GPU work for now.
Unassigning myself so if any other contributor wants to give it a go, they can.
@aeon-actions-bot unassign @Adityakushwaha2006

**Comment by Adityakushwaha2006** [comment]
@aeon-actions-bot assign @Adityakushwaha2006 
With the new PR #3227, the tests can now be put in place finally.
